File: backend/socket.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"

	"github.com/gorilla/websocket"
	"github.com/khalidkhnz/2D-metaverse-app/backend/lib"
	authService "github.com/khalidkhnz/2D-metaverse-app/backend/services/auth"
	"github.com/khalidkhnz/2D-metaverse-app/backend/types"
)

var Upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	CheckOrigin: func(r *http.Request) bool {
		return r.Header.Get("Origin") == "http://localhost:4000" || r.Header.Get("Origin") == "http://localhost:3000" || r.Header.Get("Origin") == "http://127.0.0.1:4000" || r.Header.Get("Origin") == "http://127.0.0.1:3000"
	},
}

var SOCKET_CONNECTIONS = make(map[string]*websocket.Conn)


func WSHandler(w http.ResponseWriter, r *http.Request) {

	token := r.URL.Query().Get("token")
	if token=="" {
		log.Println("WebSocket connection failed token not found")
		return
	}

	userProfile,err := authService.GetUserFromToken(token,true)
	if err!=nil {
		log.Println("WebSocket connection failed: ",err.Error())
		return
	}

	conn, err := Upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Println("WebSocket connection to 'ws://localhost:4000/ws' failed: ", err)
		return
	}

	log.Printf("New WebSocket connection from %s\n with Auth_Id: %s\n", conn.RemoteAddr(),userProfile.User.ID.Hex())

	// Store the new connection in SOCKET_CONNECTIONS
	SOCKET_CONNECTIONS[userProfile.User.ID.Hex()] = conn

	// Emit "REMOTE-ADDR" event to the new connection
	err = conn.WriteJSON(map[string]any{
		"type":"SERVER:REMOTE-ADDR",
		"payload": map[string]any{"remoteAddr": conn.RemoteAddr().String()},
	})
	if err != nil {
		log.Println(err)
		return
	}

	// GET MESSAGE AND RESPONSE
	for {
		_, message, err := conn.ReadMessage()
		if err != nil {
			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
				log.Println("Unexpected WebSocket closure:", err)
			} else {
				log.Println("WebSocket closed:", err)
			}
			break // Exit the loop on close
		}
	
		// Try to unmarshal the message as JSON
		var jsonData map[string]any
		if err := json.Unmarshal(message, &jsonData); err == nil {
			// If message is JSON, handle it
			err = handleSocketEvents(jsonData, conn, userProfile)
			if err != nil {
				log.Printf("Error handling event from %s : %s \n", userProfile.User.FullName, err)
				// Optionally, notify the client of this specific error
				conn.WriteJSON(map[string]any{
					"type":    "ERROR",
					"payload": map[string]string{"message": err.Error()},
				})
			}
		} else {
			// If message is not JSON, send error response to client and continue
			log.Printf("Received non-JSON message from %s: %s\n", conn.RemoteAddr(), string(message))
			err = conn.WriteJSON(map[string]any{
				"type":    "ERROR",
				"payload": map[string]string{"message": "Only JSON messages are allowed"},
			})
			if err != nil {
				log.Println("Error sending error response:", err)
				break // Exit if unable to communicate with the client
			}
		}
	}
	
}

func handleSocketEvents(data map[string]any, conn *websocket.Conn, userProfile *types.FullProfile) (error) {
	log.Printf("Recv JSON from %s: \n %+v\n", userProfile.User.FullName, data)

	if data["type"] == "BROADCAST" {
		err := conn.WriteJSON(map[string]any{
			"type": "BROADCAST",
			"payload": data,
		})
		return err
	}

	if data["type"] == "DISCONNECT" {
		RemoveSocketConnection(userProfile.User.ID.Hex())
	}
	
	return fmt.Errorf("invalid event")
}

func RemoveSocketConnection(id string) {
    delete(SOCKET_CONNECTIONS, id)
}

func EmitEventTo(socketIds []string, eventType string, payload map[string]any) {
	for socketId, conn := range SOCKET_CONNECTIONS {
		if lib.Contains(socketIds, socketId) {
			err := conn.WriteJSON(map[string]any{
				"type":    eventType,
				"payload": payload,
			})
			log.Println(err.Error())
		}
	}
}





// Below is the JavaScript code to connect to the WebSocket server
///////////////////////
// const socket = new WebSocket('ws://localhost:4000/ws');

// socket.onopen = function(event) {
//     console.log('WebSocket connection established');

//     // Create a JSON object to send
//     const jsonData = {
//         message: "Hello Server!",
//         timestamp: new Date().toISOString()
//     };

//     // Send the JSON object as a string
//     socket.send(JSON.stringify(jsonData));
// };

// socket.onmessage = function(event) {
//     try {
//         // Attempt to parse the server response as JSON
//         const data = JSON.parse(event.data);
//         console.log('JSON Message from server:', data);
//     } catch (e) {
//         // If parsing fails, log the message as a string
//         console.log('Message from server:', event.data);
//     }
// };

// socket.onerror = function(event) {
//     console.error('WebSocket error:', event);
// };

// socket.onclose = function(event) {
//     console.log('WebSocket connection closed');
// };
///////////////////////

/*
const socket = new WebSocket('ws://localhost:4000/ws');

socket.onopen = function(event) {
    console.log('WebSocket connection established');
    socket.send('Hello Server!');
};

socket.onmessage = function(event) {
    console.log('Message from server: ', event.data);
};

socket.onerror = function(event) {
    console.error('WebSocket error: ', event);
};

socket.onclose = function(event) {
    console.log('WebSocket connection closed');
};
*/

File: frontend/services/socket.ts
"use client";

import { wait } from "@/lib/utils";
import { getMethod } from "./ApiInterceptor";
import { ENDPOINTS } from "@/lib/Endpoints";

type ClientEvents =
  | "CLIENT:REMOTE-ADDR"
  | "CLIENT:POSITION"
  | "CLIENT:SEND-MESSAGE";
type ServerEvents =
  | "SERVER:REMOTE-ADDR"
  | "SERVER:LOBBY-POSITIONS"
  | "SERVER:RECV-MESSAGE";

class WS {
  public socket!: WebSocket;
  private retryAttempts = 0;
  private maxRetries = Infinity;
  private retryDelay = 5000;
  private externalMessageHandlers: ((event: MessageEvent<any>) => void)[] = [];

  public setupSocketConnection() {
    this.connectSocket();
  }

  private async getWebsocketToken(): Promise<string> {
    const response = await getMethod<{ success: boolean; token: string }>(
      ENDPOINTS.AUTH.GENERATE_WS_TOKEN,
    ).catch((err) => console.log(err));
    if (response?.success) {
      return response?.token;
    }
    return "";
  }

  private async connectSocket() {
    if (this.socket && this.socket.readyState === WebSocket.OPEN) {
      console.log("WebSocket is already connected.");
      return;
    }

    const token = localStorage.getItem("token")
      ? JSON.parse(localStorage.getItem("token") || "")
      : null;

    if (!token?.value) {
      this.retryConnection();
      return;
    }

    const wsToken = await this.getWebsocketToken();

    if (!wsToken) {
      this.retryConnection();
      return;
    }

    this.socket = new WebSocket(
      `ws://localhost:4000/ws${wsToken ? `?token=${wsToken}` : ""}`,
    );
    this.socketConnectionEvents({
      handleSocketEvents: this.handleSocketEvents,
      onConnectSend: { type: "CONNECT", payload: { userId: "123123" } },
    });
  }

  private async retryConnection() {
    if (this.retryAttempts < this.maxRetries) {
      this.retryAttempts++;
      console.log(`Attempting to reconnect... (Attempt ${this.retryAttempts})`);
      // Toast.info(`Attempting to reconnect... (Attempt ${this.retryAttempts})`);
      await wait(this.retryDelay);
      this.connectSocket();
    } else {
      console.error("Max retry attempts reached. Could not reconnect.");
      // Toast.error("Unable to reconnect to the server.");
    }
  }

  private socketConnectionEvents = async ({
    handleSocketEvents,
    onConnectSend,
  }: {
    handleSocketEvents: (event: MessageEvent<any>) => void;
    onConnectSend: any;
  }) => {
    this.socket.onopen = async () => {
      console.log("WebSocket connection established");
      this.retryAttempts = 0;
      await wait(1000);
      this.socket.send(JSON.stringify(onConnectSend));
    };

    this.socket.onmessage = (event) => {
      handleSocketEvents(event);
      this.externalMessageHandlers.forEach((handler) => handler(event));
    };

    this.socket.onerror = () => {
      console.log("WebSocket error occurred");
      this.retryConnection();
    };

    this.socket.onclose = () => {
      console.log("WebSocket connection closed");
      this.retryConnection();
    };
  };

  private handleSocketEvents(e: MessageEvent<any>) {
    const data: { type: ServerEvents; payload: any } = JSON.parse(e.data);
    console.log("JSON from server:", data);

    switch (data?.type) {
      case "SERVER:REMOTE-ADDR": {
        const payload: { remoteAddr: string } = data.payload;
        console.log("REMOTE-ADDR", payload);
        break;
      }
      case "SERVER:RECV-MESSAGE": {
        const payload = data.payload;
        console.log("SERVER:RECV-MESSAGE", payload);
        break;
      }
      case "SERVER:LOBBY-POSITIONS": {
        const payload = data.payload;
        console.log("SERVER:LOBBY-POSITIONS", payload);
        break;
      }
    }
  }

  public sendSocketEvent({
    type,
    payload,
  }: {
    type: ClientEvents;
    payload: any;
  }) {
    this.socket?.send(JSON.stringify({ type, payload }));
  }

  public addMessageHandler(handler: (event: MessageEvent<any>) => void) {
    this.externalMessageHandlers.push(handler);
  }

  public removeMessageHandler(handler: (event: MessageEvent<any>) => void) {
    this.externalMessageHandlers = this.externalMessageHandlers.filter(
      (h) => h !== handler,
    );
  }
}

export const socketService = new WS();
